app/project_type/git.py

```diff
@@ -128,6 +109,26 @@ def __init__(self, url, build_project_directory='', project_directory='', remote
         os.symlink(actual_project_directory, build_project_directory)
         self.project_directory = build_project_directory
 
+    def slave_param_overrides(self):
+        """
+        Produce a set of values to override original project type params for use on a slave machine.
+
+        :return: A set of values to override original project type params
+        :rtype: dict[str, str]
+        """
+        param_overrides = super().slave_param_overrides()
+
+        # We modify the repo url so the slave clones or fetches from the master directly. This should be faster than
+        # cloning/fetching from the original git remote.
+        master_repo_url = 'ssh://{}{}'.format(Configuration['hostname'], self._repo_directory)
+        param_overrides['url'] = master_repo_url  # This causes the slave to clone directly from the master.
+
+        # The user-specified branch is overwritten with a locally created ref so that slaves working on a job can
+        # continue to fetch the same HEAD, even if the master resets the user-specified branch for another build.
+        param_overrides['branch'] = self._branch
```

Comment:
It took a little while wrapping my head around the order in which the local_ref gets created in the first place (on the master upon build preparation) and when that gets propagated to the slave.

One of the worries I had about this line is that it may not be clear whether this._branch has already been properly populated. There is nothing we can do about getting the local ref value, however I think it'd be slightly easier to understand and not screw-up the ordering of (in future PR's from other developers) if you had a separate local variable called self._local_ref.

Maybe in line 163 you can set self._local_ref instead of a temporary variable, and not overwrite self._branch at all. And then in this line (line 128) you could set param_overrides['branch'] = self._local_ref. This way it's really obvious that the slave_param_overrides is taking the 'branch' argument from self._local_ref of the master and it's really obvious where that gets set (self._fetch_project).